barretenberg/cpp/src/barretenberg/commitment_schemes/gemini/gemini.hpp

```diff
@@ -100,14 +100,134 @@ template <typename Curve> class GeminiProver_ {
+        bool has_random_polynomial = false;
+
+        RefVector<Polynomial> unshifted;            // set of unshifted polynomials
+        RefVector<Polynomial> to_be_shifted_by_one; // set of polynomials to be left shifted by 1
+
+        Polynomial batched_unshifted;            // linear combination of unshifted polynomials
+        Polynomial batched_to_be_shifted_by_one; // linear combination of to-be-shifted polynomials
+
+      public:
+        PolynomialBatcher(const size_t full_batched_size)
+            : full_batched_size(full_batched_size)
+            , batched_unshifted(full_batched_size)
+            , batched_to_be_shifted_by_one(Polynomial::shiftable(full_batched_size))
+        {}
+
+        bool has_unshifted() const { return unshifted.size() > 0; }
+        bool has_to_be_shifted_by_one() const { return to_be_shifted_by_one.size() > 0; }
+
+        // Set references to the polynomials to be batched
+        void set_unshifted(RefVector<Polynomial> polynomials) { unshifted = polynomials; }
+        void set_to_be_shifted_by_one(RefVector<Polynomial> polynomials) { to_be_shifted_by_one = polynomials; }
+
+        // Initialize the random polynomial used to add randomness to the batched polynomials for ZK
+        void set_random_polynomial(Polynomial&& random)
+        {
+            has_random_polynomial = true;
+            random_polynomial = random;
+        }
+
+        /**
+         * @brief Compute batched polynomial A₀ = F + G/X as the linear combination of all polynomials to be opened
+         * @details If the random polynomial is set, it is added to the batched polynomial for ZK
+         *
+         * @param challenge batching challenge
+         * @param running_scalar power of the batching challenge
+         * @return Polynomial A₀
+         */
+        Polynomial compute_batched(const Fr& challenge, Fr& running_scalar)
+        {
+            Polynomial full_batched(full_batched_size);
+
+            // lambda for batching polynomials; updates the running scalar in place
+            auto batch = [&](Polynomial& batched, const RefVector<Polynomial>& polynomials_to_batch) {
+                for (auto& poly : polynomials_to_batch) {
+                    batched.add_scaled(poly, running_scalar);
+                    running_scalar *= challenge;
+                }
+            };
+
+            // if necessary, add randomness to the full batched polynomial for ZK
+            if (has_random_polynomial) {
+                full_batched += random_polynomial;
+            }
+
+            // compute the linear combination F of the unshifted polynomials
+            if (has_unshifted()) {
+                batch(batched_unshifted, unshifted);
+                full_batched += batched_unshifted; // A₀ = F
+            }
+
+            // compute the linear combination G of the to-be-shifted polynomials
+            if (has_to_be_shifted_by_one()) {
+                batch(batched_to_be_shifted_by_one, to_be_shifted_by_one);
```

Comment:
This pattern is slightly less nice than `batched = batch(polys)` but I need to initialize the `batched_unshifted` and `batched_to_be_shifted_by_one` prior to this to ensure they're correctly sized. I thought about initializing within the method based on e.g. the size of the first poly in the set to be batched but there's no guarantee that all of the polys have the "full" size.